src/dlfcn.c

```diff
@@ -474,7 +479,202 @@ char *dlerror( void )
+        IMAGE_THUNK_DATA *thunkILT = (IMAGE_THUNK_DATA*)(iid[i].Characteristics + base);
+        IMAGE_THUNK_DATA *thunkIAT = (IMAGE_THUNK_DATA*)(iid[i].FirstThunk + base);
+        for(; thunkILT->u1.AddressOfData != 0; thunkILT++, thunkIAT++) {
+            IMAGE_IMPORT_BY_NAME *nameData;
+            if (IMAGE_SNAP_BY_ORDINAL(thunkILT->u1.Ordinal))
+              continue;
+            if( (void *)thunkIAT->u1.Function <= addr && candidateAddr < (void *)thunkIAT->u1.Function ) {
+                candidateAddr = (void *)thunkIAT->u1.Function;
+                nameData = (IMAGE_IMPORT_BY_NAME*)( thunkILT->u1.AddressOfData + base );
+                candidateName = (const char *)nameData->Name;
+            }
+        }
+   }
+   *func_address = candidateAddr;
+   return candidateName;
+}
+
+static BOOL is_valid_address( void *addr )
+{
+    if( addr == NULL )
+        return FALSE;
+
+    /* check valid pointer */
+    MEMORY_BASIC_INFORMATION info;
+    SIZE_T result;
+    result = VirtualQuery( addr, &info, sizeof( info ) );
+    if( result == 0 || info.AllocationBase == NULL || info.AllocationProtect == 0 || info.AllocationProtect == PAGE_NOACCESS )
+        return FALSE;
+    return TRUE;
+}
+
+/*
+ * Return state if address points to an import thunk
+ *
+ * An import thunk is setup with a 'jmp' instruction followed by an
+ * absolute address (32bit) or relative offset (64bit) pointing into
+ * the import address table (iat), which is partially maintained by
+ * the runtime linker.
+ */
+static BOOL is_import_thunk( void *addr )
+{
+    return *(short *)addr == 0x25ff ? TRUE : FALSE;
+}
+
+/*
+ * Return adress from the import address table (iat), if
+ * the original address points to a thunk table entry.
+ */
+static void *get_address_from_import_address_table( void *iat, DWORD iat_size, void *addr )
+{
+    BYTE *thkp = (BYTE*)addr;
+    /* get offset from thunk table (after instruction 0xff 0x25)
+     *   4018c8 <_VirtualQuery>: ff 25 4a 8a 00 00
+     */
+    ULONG offset = *(ULONG*)(thkp + 2);
+#ifdef _WIN64
+    /* On 64 bit the offset is relative
+     *   4018c8:	ff 25 4a 8a 00 00    	jmpq   *0x8a4a(%rip)        # 40a318 <__imp_VirtualQuery>
+     */
+    BYTE *ptr = (BYTE *)(thkp + 6 + offset);
+#else
+    /* On 32 bit the offset is absolute
+     *    4019b4:	ff 25 90 71 40 00    	jmp    *0x40719
+     */
+    BYTE *ptr = (BYTE *)offset;
+#endif
+
+    if( !is_valid_address( ptr ) ||  ptr < (BYTE *)iat || ptr > (BYTE *)iat + iat_size )
+        return NULL;
+
+    return *(void **)ptr;
+}
+
+/* holds module filename */
+static char module_filename[2*MAX_PATH];
+
+static BOOL fill_module_info( void *addr, Dl_info *info )
+{
+    HMODULE hModule;
+    DWORD dwSize;
+
+    info->dli_fname = NULL;
+    if( !GetModuleHandleExA( GET_MODULE_HANDLE_EX_FLAG_FROM_ADDRESS | GET_MODULE_HANDLE_EX_FLAG_UNCHANGED_REFCOUNT, addr, &hModule ) || !hModule )
+    {
+        info->dli_fbase = NULL;

```

Comment:
When `fill_module_info()` returns FALSE then `dladdr()` return 0. So No need to fill `dli_fbase` in this case.

Reply:
Thanks for pointing out - fixed on other locations too.